uefi-raw/src/protocol/network/tcpv4.rs

```diff
@@ -0,0 +1,513 @@
+
+    /// Add or delete routing entries.
+    ///
+    /// The most specific route is selected by comparing the
+    /// `subnet_address` with the destination IP address
+    /// arithmetically ANDed with the `subnet_mask`.
+    ///
+    /// The default route is added with both `subnet_address` and
+    /// `subnet_mask` set to `0.0.0.0`. The default route matches all
+    /// destination IP addresses if there is no more specific route.
+    ///
+    /// A direct route is added with `gateway_address` set to
+    /// `0.0.0.0`. Packets are sent to the destination host if its
+    /// address can be found in the Address Resolution Protocol (ARP)
+    /// cache or it is on the local subnet. If the instance is
+    /// configured to use default address, a direct route to the local
+    /// network will be added automatically.
+    ///
+    /// Each TCP instance has its own independent routing table. An
+    /// instance that uses the default IP address will have a copy of
+    /// the
+    /// [`Ipv4Config2Protocol`][super::ip4_config2::Ip4Config2Protocol]'s
+    /// routing table. The copy will be updated automatically whenever
+    /// the IP driver reconfigures its instance. As a result, the
+    /// previous modification to the instance's local copy will be
+    /// lost.
+    ///
+    /// The priority of checking the route table is specific to the IP
+    /// implementation, and every IP implementation must comply with
+    /// RFC 1122.
+    ///
+    /// Note: There is no way to set up routes to other network
+    /// interface cards (NICs) because each NIC has its own
+    /// independent network stack that shares information only through
+    /// EFI TCPv4 variable.
+    pub routes: unsafe extern "efiapi" fn(
+        this: *mut Self,
+        delete_route: Boolean,
+        subnet_address: *const Ipv4Address,
+        subnet_mask: *const Ipv4Address,
+        gateway_address: *const Ipv4Address,
+    ) -> Status,
+
+    /// Initiate a nonblocking TCP connection request for an active
+    /// TCP instance.
+    ///
+    /// `connect` initiates an active open to the remote peer
+    /// configured in the current TCP instance if it is configured as
+    /// active. If the connection succeeds or fails due to any error,
+    /// the [`connection_token.event`][Tcpv4CompletionToken::event]
+    /// will be signaled and `connection_token.status` will be updated
+    /// accordingly. This function can only be called for the TCP
+    /// instance in the [`Tcpv4ConnectionState::CLOSED`] state. The
+    /// instance will transition to [`Tcpv4ConnectionState::SYN_SENT`]
+    /// if the function returns [`Status::SUCCESS`]. If the TCP
+    /// three-way handshake succeeds, its state will become
+    /// [`Tcpv4ConnectionState::ESTABLISHED`], otherwise, the state
+    /// will return to [`Tcpv4ConnectionState::CLOSED`].
+    pub connect: unsafe extern "efiapi" fn(
+        this: *mut Self,
+        connection_token: *mut Tcpv4CompletionToken,
```

Comment:
This should be a ConnectionToken rather than a CompletionToken (even though they are essentially the same type)

Reply:
for the record. `EFI_TCP4_CONNECTION_TOKEN` and `EFI_TCP4_COMPLETION_TOKEN` have the same structure.

https://uefi.org/specs/UEFI/2.11/28_Network_Protocols_TCP_IP_and_Configuration.html#efi-tcp4-protocol-connect